tpm2/policy.go

```diff
@@ -151,189 +152,263 @@ func makeStaticPolicyDataRaw_v1(data *staticPolicyData) *staticPolicyDataRaw_v1
+func (c *pcrPolicyCounterCommon) increment(tpm *tpm2.TPMContext, from, to uint64, key crypto.PrivateKey, cb func(tpm2.SessionContext) error, session tpm2.SessionContext) error {
+	if from > to {
+		return errors.New("counter is already higher then desired value")
+	}
+
+	index, err := tpm2.CreateNVIndexResourceContextFromPublic(c.pub)
 	if err != nil {
 		return xerrors.Errorf("cannot create context for NV index: %w", err)
 	}
 
 	// Begin a policy session to increment the index.
-	policySession, err := tpm.StartAuthSession(nil, nil, tpm2.SessionTypePolicy, nil, nvPublic.NameAlg)
+	policySession, err := tpm.StartAuthSession(nil, nil, tpm2.SessionTypePolicy, nil, c.pub.NameAlg)
 	if err != nil {
 		return xerrors.Errorf("cannot begin policy session: %w", err)
 	}
 	defer tpm.FlushContext(policySession)
 
-	// Compute a digest for signing with the update key
 	signDigest := tpm2.HashAlgorithmNull
-	keyScheme := keyPublic.Params.AsymDetail().Scheme
+	keyScheme := c.updateKey.Params.AsymDetail().Scheme
 	if keyScheme.Scheme != tpm2.AsymSchemeNull {
 		signDigest = keyScheme.Details.Any().HashAlg
 	}
 	if signDigest == tpm2.HashAlgorithmNull {
 		signDigest = tpm2.HashAlgorithmSHA256
 	}
-	h := signDigest.NewHash()
-	h.Write(policySession.NonceTPM())
-	binary.Write(h, binary.BigEndian, int32(0)) // expiration
-
-	// Sign the digest
-	var signature tpm2.Signature
-	switch k := key.(type) {
-	case *rsa.PrivateKey:
-		sig, err := rsa.SignPSS(rand.Reader, k, signDigest.GetHash(), h.Sum(nil), &rsa.PSSOptions{SaltLength: rsa.PSSSaltLengthEqualsHash})
-		if err != nil {
-			return xerrors.Errorf("cannot sign authorization: %w", err)
-		}
-		signature = tpm2.Signature{
-			SigAlg: tpm2.SigSchemeAlgRSAPSS,
-			Signature: &tpm2.SignatureU{
-				RSAPSS: &tpm2.SignatureRSAPSS{
-					Hash: signDigest,
-					Sig:  tpm2.PublicKeyRSA(sig)}}}
-	case *ecdsa.PrivateKey:
-		sigR, sigS, err := ecdsa.Sign(rand.Reader, k, h.Sum(nil))
-		if err != nil {
-			return xerrors.Errorf("cannot sign authorization: %w", err)
-		}
-		signature = tpm2.Signature{
-			SigAlg: tpm2.SigSchemeAlgECDSA,
-			Signature: &tpm2.SignatureU{
-				ECDSA: &tpm2.SignatureECDSA{
-					Hash:       signDigest,
-					SignatureR: sigR.Bytes(),
-					SignatureS: sigS.Bytes()}}}
-	default:
-		panic("invalid private key type")
-	}
 
 	// Load the public part of the key in to the TPM. There's no integrity protection for this command as if it's altered in
 	// transit then either the signature verification fails or the policy digest will not match the one associated with the NV
 	// index.
-	keyLoaded, err := tpm.LoadExternal(nil, keyPublic, tpm2.HandleEndorsement)
+	keyLoaded, err := tpm.LoadExternal(nil, c.updateKey, tpm2.HandleEndorsement)
 	if err != nil {
 		return xerrors.Errorf("cannot load public part of key used to verify authorization signature: %w", err)
 	}
 	defer tpm.FlushContext(keyLoaded)
 
-	// Execute the policy assertions
-	if version == 0 {
```

Comment:
this if had very asymmetric sides to start with but the new code is even harder to follow because the logic is spread between a callback that is used only in one case and fixed field of the common struct etc.

Reply:
I've refactored this now. Whilst doing this, I also added a helper API to go-tpm2 for creating the signed assertions used by this code (https://pkg.go.dev/github.com/canonical/go-tpm2/util@master#SignPolicyAuthorization)